<?xml version="1.0" encoding="utf-8"?>
<launch>
    <!-- Mode selection: dummy=true for visualization-only, dummy=false for full Gazebo + WBC -->
    <arg name="dummy" default="false"/>

    <!-- Required arguments -->
    <arg name="controllers_ns"      default="franka"/>
    <arg name="controllers_args"    default="joint_controller"/>
    <arg name="tbai_config_path"    default="$(find tbai_ros_mpc)/config/franka/default_config_franka.yaml"/>
    <arg name="description_name"    default="robot_description"/>
    <arg name="description_file"    default="$(find tbai_ros_description)/urdf/franka_gazebo.urdf.xacro"/>

    <!-- MPC/WBC specific paths -->
    <arg name="taskFile"  default="$(find tbai_ros_mpc)/../dependencies/tbai/tbai_mpc/config/franka/task.info"/>
    <arg name="urdfFile"  default="$(find gazebo_robotic_assets)/resources/mobile_manipulator/franka/urdf/panda.urdf"/>
    <arg name="libFolder" default="/tmp/tbai_franka_wbc"/>

    <!-- Gazebo parameters (only used when dummy=false) -->
    <arg name="gui" default="false"/>
    <arg name="paused" default="false"/>
    <arg name="world" default="normal"/>
    <arg name="x" default="0.0"/>
    <arg name="y" default="0.0"/>
    <arg name="z" default="0.5"/>

    <!-- Debug and visualization -->
    <arg name="debug" default="false"/>
    <arg name="rviz" default="true"/>

    <!-- Upload config paths to ROS param server -->
    <param name="tbai_config_path" type="string" value="$(arg tbai_config_path)"/>
    <param name="taskFile" type="string" value="$(arg taskFile)"/>
    <param name="urdfFile" type="string" value="$(arg urdfFile)"/>
    <param name="libFolder" type="string" value="$(arg libFolder)"/>

    <!-- Set environment variable for config path -->
    <env name="TBAI_GLOBAL_CONFIG_PATH" value="$(arg tbai_config_path)"/>

    <group if="$(arg dummy)">
        <!-- Load robot description -->
        <param name="robot_description" textfile="$(arg urdfFile)"/>

        <!-- MPC node -->
        <node if="$(arg debug)" pkg="tbai_ros_mpc" type="franka_mpc_node" name="franka_mpc_node"
              output="screen" launch-prefix="gnome-terminal -- gdb -ex run --args"/>
        <node unless="$(arg debug)" pkg="tbai_ros_mpc" type="franka_mpc_node" name="franka_mpc_node"
              output="screen"/>

        <!-- Dummy MRT node -->
        <node pkg="tbai_ros_mpc" type="franka_dummy_mrt_node" name="franka_dummy_mrt_node"
              output="screen"/>

        <!-- Target interactive marker -->
        <node if="$(arg rviz)" pkg="tbai_ros_mpc" type="franka_target" name="franka_target"
              output="screen"/>
    </group>

    <group unless="$(arg dummy)">
        <!-- Launch gazebo -->
        <include file="$(find tbai_ros_gazebo)/launch/world.launch">
            <arg name="controllers_ns" value="$(arg controllers_ns)"/>
            <arg name="controllers_args" value="$(arg controllers_args)"/>
            <arg name="config_file" value="$(arg tbai_config_path)"/>
            <arg name="description_name" value="$(arg description_name)"/>
            <arg name="description_file" value="$(arg description_file)"/>
            <arg name="gui" value="$(arg gui)"/>
            <arg name="world" value="$(arg world)"/>
            <arg name="x" value="$(arg x)"/>
            <arg name="y" value="$(arg y)"/>
            <arg name="z" value="$(arg z)"/>
            <arg name="paused" value="$(arg paused)"/>
        </include>

        <!-- Launch Franka WBC node -->
        <node if="$(arg debug)" name="franka_wbc" pkg="tbai_ros_mpc" type="simple_franka_wbc" output="screen"
              launch-prefix="gnome-terminal -- gdb -ex run --args"/>
        <node unless="$(arg debug)" name="franka_wbc" pkg="tbai_ros_mpc" type="simple_franka_wbc" output="screen"/>

        <!-- Launch virtual joystick -->
        <node name="virtual_joystick" pkg="tbai_ros_utils" type="virtual_joystick.py" output="screen"/>
    </group>

    <group if="$(arg rviz)">
        <arg name="rvizconfig" default="$(find tbai_ros_mpc)/rviz/franka.rviz"/>
        <node pkg="rviz" type="rviz" name="rviz" args="-d $(arg rvizconfig)" output="screen"/>
    </group>

</launch>
